Initial support for array connectors

pmai opened the pull request:
Dimension elements on connectors. More has to be added to specify connection semantics, etc.

Also needs corresponding changes in the SSV/SSM/... side of things.

pmai commented:
Connection semantics and SSV/SSM/... changes to be handled in a separate PR.

klausschuch commented on docs/5___ssd.adoc:
```diff
@@ -126,11 +126,37 @@ The following XML child elements are specified for the Connector element:
+
+This optional element specifies one dimension of an array connector.
+If no dimension elements are present in a connector, it is a scalar connector.
+The number of dimension elements in a connector provides the dimensionality of the array.
```
The number of dimension elements in a connector defines the dimensionality of the array.

klausschuch commented on docs/5___ssd.adoc:
```diff
@@ -126,11 +126,37 @@ The following XML child elements are specified for the Connector element:
+                                    
+Either the fixedSize or the sizeConnector attributes *CAN* be present on the element, indicating a fixed size, or a size that depends on the structural parameter or constant referenced by the sizeConnector attribute.
+If none of the attributes are present, then the size of the dimension is unspecified at the SSD level.
+If both attributes are present this is considered an error.
```
If both attributes are present, this is considered an error.

klausschuch commented on schema/SystemStructureDescription.xsd:
```diff
@@ -602,6 +602,45 @@
+                                <xs:documentation xml:lang="en">
+                                    This optional element specifies one dimension of an array connector.
+                                    If no dimension elements are present in a connector, it is a scalar
+                                    connector. The number of dimension elements in a connector provides
```
defines

DagBruck replied:
@klausschuch can you explain why you prefer `defines`?

klausschuch commented on docs/5___ssd.adoc:
```diff
@@ -126,11 +126,37 @@ The following XML child elements are specified for the Connector element:
+If no dimension elements are present in a connector, it is a scalar connector.
+The number of dimension elements in a connector provides the dimensionality of the array.
+                                    
+Either the fixedSize or the sizeConnector attributes *CAN* be present on the element, indicating a fixed size, or a size that depends on the structural parameter or constant referenced by the sizeConnector attribute.
```
We might want to avoid "fixed" as part of "fixedSize" since fixed has a certain meaning in the FMI standard. I suggest to call this attribute "constantSize" or just "size".

DagBruck replied:
I agree with @klausschuch here.

pmai replied:
Changed to size.

DagBruck reviewed (commented):
I agree in general, but I had two comments on the feedback from @klausschuch.

pmai commented:
I have moved the dimension elements into common, and added array support for SSV and SSB (SSM should not currently be affected, unless we want to enhance transformations to also provide array shape transformations, but even in this case this would likely go into common). This also adds the capability to provide array values for parameters.

I have also implemented the non-controversial subset (0-2) of the connection options of #9, by providing optional startIndices and endIndices attributes on a connection. This currently allows for:

0) Only same-shaped arrays can be connected:+
    No `startIndices` and `endIndices` specified
1) from a scalar to a single element of an array: a -> A(1,1).+
    No `startIndices`, `endIndices="1 1"`
2) from a single element of an array to a single element of an array: A(1,1) -> B(1,1).+
    `startIndices="1 1"` `endIndices="1 1"`

Case 3) (scalar mapped to all elements of an array) could still be allowed, by having no `startIndices` and no `endIndices`, and defining this to mean mapping to all elements, but I have not added verbiage to this effect. Case 4) (single element of an array to all elements of an array) could also be handled in this way.

Cases 5) and 6) could be handled by extending the allowed content of the Indices attributes to not only contain indices, but also range expressions, like `startIndices="1-5 2-3"` or even `stardIndices="3-n 4-m"` or similar. However again this is not currently part of the proposal and could be added in a later SSP release.

Another approach would be to handle this in Transformations, i.e. adding ArrayTransformations that do reshaping; this would also apply naturally to parameters, which we currently do not handle (except via mapping to parameter connectors, and then using connections). However it seems a bit overkill to always have to provide an ArrayTransformation for the simple cases 1/2, hence the current approach.

JochenKoehler commented:
Web meeting 2023-11-24
Decision made to approve the pull requrest. Pierre will do the merge.

pmai commented on docs/6___ssv.adoc:
```diff
@@ -66,13 +66,16 @@ The following XML child elements are specified for the Parameter element:
 Integer / Int8 / UInt8 / Int16 / UInt16 / +
 Int32 / UInt32 / Int64 / UInt64 / +
 Boolean / String / Enumeration / Binary |Exactly one of these elements *MUST* be present to specify the type of the parameter. See below for details on each type.
+|Dimension |One or more of these optional elements define the parameter array dimensions, making the parameter an array parameter. See <<ssc_dimensions>> for details.
```
```suggestion
|Dimension |One or more of these optional elements specify the parameter array dimensions, making the parameter an array parameter. See <<ssc_dimensions>> for details.
```

pmai commented on docs/5___ssd.adoc:
```diff
@@ -133,12 +133,15 @@ The following XML child elements are specified for the Connector element:
 Int32 / UInt32 / Int64 / UInt64 / +
-Boolean / String / Enumeration / Binary |Exactly one of these elements *CAN* be present to specify the type of the Connector. See 4.5.1 Type Choice for details.
+Boolean / String / Enumeration / Binary |Exactly one of these elements *CAN* be present to specify the type of the Connector. See <<ssc_types>> for details.
+|Dimension |One or more of these optional elements define the connector array dimensions, making the connector an array connector. See <<ssc_dimensions>> for details.
```
```suggestion
|Dimension |One or more of these optional elements specify the connector array dimensions, making the connector an array connector. See <<ssc_dimensions>> for details.
```
